### PR TITLE
Ensure that a node's CIDR isn't released until the node is deleted

### LINKED ISSUE #127792: kube-controller-manager re-allocating already used PodCIDR (v1.31)
### What happened?

Since upgrading to Kubernetes v1.31 we have occasional observed that `kube-controller-manager` allocates a PodCIDR for a new node which is already allocated for an existing node in the cluster. We have about 200 clusters with a combined number of ca. 6k-10k nodes (autoscaling). It has happened about 10 times since upgrading a week ago, so it's rather rare but breaks the networking between the affected nodes when it happens.

We use flannel as overlay network and it uses the podCIDR assigned by `kube-controller-manager`. When a new node gets an existing podCIDR assigned, then the new node doesn't get ready and it gets terminated after 10 min. We check that the node is routable by all other nodes before we mark the node ready. This never happens because the podCIDR already points to another node in the cluster. When the new node is deleted it has the very unfortunate side effect that all flannels detect the node is deleted and they delete the routes for the podCIDR, problem is just that most nodes will have routes for the other node with the same podCIDR and this then stops working.

Terminating the affected node is the best way to resolve the issue but obviously not sustainable.

### What did you expect to happen?

`kube-controller-manager` should _never_ allocate the same podCIDR to multiple nodes at the same time.

### How can we reproduce it (as minimally and precisely as possible)?

We haven't found a predictable way to reproduce it.

We run with the pod CIDR range: `10.2.0.0/15`.

### Anything else we need to know?

This is how we configure `kube-controller-manager`: https://github.com/zalando-incubator/kubernetes-on-aws/blob/dev/cluster/node-pools/master-default/userdata.yaml#L581-L648 (the image is vanilla Kubernetes built from source and hosted in a private registry)

Here is an example from one cluster with logs:

*Node 1 (original)*:
name: `ip-172-31-14-210.eu-central-1.compute.internal`
podCIDR: `10.2.4.0/24`

*Node 2 (new)*:
name: `ip-172-31-15-241.eu-central-1.compute.internal`
podCIDR: `10.2.4.0/24`

`kube-controller-manager`  logs filtered for the node `ip-172-31-14-210.eu-central-1.compute.internal`:

```
I1001 16:54:22.944453       1 actual_state_of_world.go:540] "Failed to update statusUpdateNeeded field in actual state of world" logger="persistentvolume-attach-detach-controller" err="Failed to set statusUpdateNeeded to needed true, because nodeName=\"ip-172-31-14-210.eu-central-1.compute.internal\" does not exist"
I1001 16:54:23.001587       1 range_allocator.go:241] "Successfully synced" logger="node-ipam-controller" key="ip-172-31-14-210.eu-central-1.compute.internal"
I1001 16:54:23.054817       1 node_lifecycle_controller.go:884] "Missing timestamp for Node. Assuming now as a timestamp" logger="node-lifecycle-controller" node="ip-172-31-14-210.eu-central-1.compute.internal"
I1001 16:56:52.839072       1 range_allocator.go:241] "Successfully synced" logger="node-ipam-controller" key="ip-172-31-14-210.eu-central-1.compute.internal"
I1001 17:01:57.639919       1 range_allocator.go:241] "Successfully synced" logger="node-ipam-controller" key="ip-172-31-14-210.eu-central-1.compute.internal"
I1001 17:03:00.584960       1 range_allocator.go:241] "Successfully synced" logger="node-ipam-controller" key="ip-172-31-14-210.eu-central-1.compute.internal"
I1001 17:03:19.873337       1 range_allocator.go:241] "Successfully synced" logger="node-ipam-controller" key="ip-172-31-14-210.eu-central-1.compute.internal"
I1001 17:08:26.273278       1 range_allocator.go:241] "Successfully synced" logger="node-ipam-controller" key="ip-172-31-14-210.eu-central-1.compute.internal"
I1001 17:13:32.137435       1 range_allocator.go:241] "Successfully synced" logger="node-ipam-controller" key="ip-172-31-14-210.eu-central-1.compute.internal"
I1001 17:18:37.436066       1 range_allocator.go:241] "Successfully synced" logger="node-ipam-controller" key="ip-172-31-14-210.eu-central-1.compute.internal"
I1001 17:20:50.207426       1 range_allocator.go:241] "Successfully synced" logger="node-ipam-controller" key="ip-172-31-14-210.eu-central-1.compute.internal"
I1001 17:22:06.023052       1 range_allocator.go:241] "Successfully synced" logger="node-ipam-controller" key="ip-172-31-14-210.eu-central-1.compute.internal"
I1001 17:22:06.058496       1 range_allocator.go:241] "Successfully synced" logger="node-ipam-controller" key="ip-172-31-14-210.eu-central-1.compute.internal"
I1001 17:22:11.270715       1 range_allocator.go:241] "Successfully synced" logger="node-ipam-controller" key="ip-172-31-14-210.eu-central-1.compute.internal"
I1001 17:22:11.376488       1 reconciler.go:275] "attacherDetacher.DetachVolume started" logger="persistentvolume-attach-detach-controller" node="ip-172-31-14-210.eu-central-1.compute.internal" volumeName="kubernetes.io/csi/ebs.csi.aws.com^vol-01a06d931a5cbc177"
I1001 17:22:11.386310       1 range_allocator.go:241] "Successfully synced" logger="node-ipam-controller" key="ip-172-31-14-210.eu-central-1.compute.internal"
I1001 17:22:11.392521       1 operation_generator.go:1539] Verified volume is safe to detach for volume "pvc-0ecc2568-e98d-430d-aa57-d545f7cb8c85" (UniqueName: "kubernetes.io/csi/ebs.csi.aws.com^vol-01a06d931a5cbc177") on node "ip-172-31-14-210.eu-central-1.compute.internal"
I1001 17:22:18.918976       1 operation_generator.go:437] DetachVolume.Detach succeeded for volume "pvc-0ecc2568-e98d-430d-aa57-d545f7cb8c85" (UniqueName: "kubernetes.io/csi/ebs.csi.aws.com^vol-01a06d931a5cbc177") on node "ip-172-31-14-210.eu-central-1.compute.internal"
I1001 17:23:15.637313       1 range_allocator.go:241] "Successfully synced" logger="node-ipam-controller" key="ip-172-31-14-210.eu-central-1.compute.internal"
```

`kube-controller-manager` logs filtered for the node: `ip-172-31-15-241.eu-central-1.compute.internal`:

```
I1001 17:22:45.543797       1 actual_state_of_world.go:540] "Failed to update statusUpdateNeeded field in actual state of world" logger="persistentvolume-attach-detach-controller" err="Failed to set statusUpdateNeeded to needed true, because nodeName=\"ip-172-31-15-241.eu-central-1.compute.internal\" does not exist"
I1001 17:22:45.647668       1 range_allocator.go:422] "Set node PodCIDR" logger="node-ipam-controller" node="ip-172-31-15-241.eu-central-1.compute.internal" podCIDRs=["10.2.4.0/24"]
I1001 17:22:45.648019       1 range_allocator.go:241] "Successfully synced" logger="node-ipam-controller" key="ip-172-31-15-241.eu-central-1.compute.internal"
I1001 17:22:45.648088       1 range_allocator.go:241] "Successfully synced" logger="node-ipam-controller" key="ip-172-31-15-241.eu-central-1.compute.internal"
I1001 17:22:45.695174       1 range_allocator.go:241] "Successfully synced" logger="node-ipam-controller" key="ip-172-31-15-241.eu-central-1.compute.internal"
I1001 17:22:45.702231       1 range_allocator.go:241] "Successfully synced" logger="node-ipam-controller" key="ip-172-31-15-241.eu-central-1.compute.internal"
I1001 17:22:45.724468       1 range_allocator.go:241] "Successfully synced" logger="node-ipam-controller" key="ip-172-31-15-241.eu-central-1.compute.internal"
I1001 17:22:45.752257       1 range_allocator.go:241] "Successfully synced" logger="node-ipam-controller" key="ip-172-31-15-241.eu-central-1.compute.internal"
I1001 17:22:45.820108       1 range_allocator.go:241] "Successfully synced" logger="node-ipam-controller" key="ip-172-31-15-241.eu-central-1.compute.internal"
I1001 17:22:45.846131       1 range_allocator.go:241] "Successfully synced" logger="node-ipam-controller" key="ip-172-31-15-241.eu-central-1.compute.internal"
I1001 17:22:48.640513       1 node_lifecycle_controller.go:884] "Missing timestamp for Node. Assuming now as a timestamp" logger="node-lifecycle-controller" node="ip-172-31-15-241.eu-central-1.compute.internal"
I1001 17:23:02.216555       1 range_allocator.go:241] "Successfully synced" logger="node-ipam-controller" key="ip-172-31-15-241.eu-central-1.compute.internal"
I1001 17:23:02.273556       1 range_allocator.go:241] "Successfully synced" logger="node-ipam-controller" key="ip-172-31-15-241.eu-central-1.compute.internal"
I1001 17:23:16.215529       1 range_allocator.go:241] "Successfully synced" logger="node-ipam-controller" key="ip-172-31-15-241.eu-central-1.compute.internal"
I1001 17:23:21.763576       1 range_allocator.go:241] "Successfully synced" logger="node-ipam-controller" key="ip-172-31-15-241.eu-central-1.compute.internal"
I1001 17:23:46.683733       1 range_allocator.go:241] "Successfully synced" logger="node-ipam-controller" key="ip-172-31-15-241.eu-central-1.compute.internal"
I1001 17:25:05.756696       1 range_allocator.go:241] "Successfully synced" logger="node-ipam-controller" key="ip-172-31-15-241.eu-central-1.compute.internal"
I1001 17:25:05.786234       1 range_allocator.go:241] "Successfully synced" logger="node-ipam-controller" key="ip-172-31-15-241.eu-central-1.compute.internal"
```

The interesting lines are:

```
I1001 16:54:23.001587       1 range_allocator.go:241] "Successfully synced" logger="node-ipam-controller" key="ip-172-31-14-210.eu-central-1.compute.internal"
...
I1001 17:22:45.647668       1 range_allocator.go:422] "Set node PodCIDR" logger="node-ipam-controller" node="ip-172-31-15-241.eu-central-1.compute.internal" podCIDRs=["10.2.4.0/24"]
```

At `16:54` the original node is `Successfully synced` by the `node-ipam-controller` and at `17:22`, much later, does the new node get the same podCIDR allocated as what is already allocated to the original node.

We can also see this from the flannel logs on the new node:

```
 time="2024-10-01T17:22:58Z" level=info msg="Added new node ip-172-31-14-210.eu-central-1.compute.internal with PodCIDR 10.2.4.0/24, address 10.2.4.0"
```

### Kubernetes version

<details>

```console
$ kubectl version
Client Version: v1.31.0
Kustomize Version: v5.4.2
Server Version: v1.31.0
```

</details>

### Cloud provider

<details>
AWS running a custom Kubernetes setup on EC2
</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

### PR DESCRIPTION
**Pull request by adrianmoisey**
Fixes https://github.com/kubernetes/kubernetes/issues/127792

Fixes bug where a node's PodCIDR was released when the node was given a delete time stamp, but was hanging around due to a finalizer.

#### What type of PR is this?

/kind bug
/kind regression

#### What this PR does / why we need it:
Fix regression where a PodCIDR is released early, before a node is deleted.

#### Which issue(s) this PR fixes:

Fixes #127792

#### Special notes for your reviewer:

I struggled a bunch with the tests. I feel like they definitely need to be improved. 

#### Does this PR introduce a user-facing change?
```release-note
Fixes a 1.31 regression where PodCIDR was released before node was deleted
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: pkg/controller/nodeipam/ipam/range_allocator_test.go [hunk 1]
```diff
@@ -841,3 +841,107 @@ func TestReleaseCIDRSuccess(t *testing.T) {
...
+
+			fakeNodeInformer := test.FakeNodeInformer(fakeNodeHandler)
+			nodeList, _ := fakeNodeHandler.List(tCtx, metav1.ListOptions{})
+			allocator, err := NewCIDRRangeAllocator(tCtx, fakeNodeHandler, fakeNodeInformer, allocatorParams, nodeList)
+			if err != nil {
+				t.Fatalf("failed to create CIDRRangeAllocator: %v", err)
+			}
+			rangeAllocator, ok := allocator.(*rangeAllocator)
+			if !ok {
+				t.Fatalf("found non-default implementation of CIDRAllocator")
```
Comment (aojea) [inline pkg/controller/nodeipam/ipam/range_allocator_test.go:918]:
```suggestion
				t.Fatalf("found non-default implementation of rangeAllocator")
```

**Comment by aojea** [comment]
@adrianmoisey  I realized that we have to change something on the handlers too

On the test I know that you took something from the existing ones, but it is better to always check all the errors, also I reorganized a bit the test to make it more clear how we are asserting 

```diff
diff --git a/pkg/controller/nodeipam/ipam/range_allocator.go b/pkg/controller/nodeipam/ipam/range_allocator.go
index 97ac56b94fd..6e8fa2e4e95 100644
--- a/pkg/controller/nodeipam/ipam/range_allocator.go
+++ b/pkg/controller/nodeipam/ipam/range_allocator.go
@@ -145,16 +145,21 @@ func NewCIDRRangeAllocator(ctx context.Context, client clientset.Interface, node
                DeleteFunc: func(obj interface{}) {
                        // The informer cache no longer has the object, and since Node doesn't have a finalizer,
                        // we don't see the Update with DeletionTimestamp != 0.
-                       // TODO: instead of executing the operation directly in the handler, build a small cache with key node.Name
-                       // and value PodCIDRs use ReleaseCIDR on the reconcile loop so we can retry on `ReleaseCIDR` failures.
-                       if err := ra.ReleaseCIDR(logger, obj.(*v1.Node)); err != nil {
-                               utilruntime.HandleError(fmt.Errorf("error while processing CIDR Release: %w", err))
+                       node, ok := obj.(*v1.Node)
+                       if !ok {
+                               tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+                               if !ok {
+                                       utilruntime.HandleError(fmt.Errorf("unexpected object type: %v", obj))
+                                       return
+                               }
+                               node, ok = tombstone.Obj.(*v1.Node)
+                               if !ok {
+                                       utilruntime.HandleError(fmt.Errorf("unexpected object types: %v", obj))
+                                       return
+                               }
                        }
-                       // IndexerInformer uses a delta nodeQueue, therefore for deletes we have to use this
-                       // key function.
-                       key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
-                       if err == nil {
-                               ra.queue.Add(key)
+                       if err := ra.ReleaseCIDR(logger, node); err != nil {
+                               utilruntime.HandleError(fmt.Errorf("error while processing CIDR Release: %w", err))
                        }
                },
        })
diff --git a/pkg/controller/nodeipam/ipam/range_allocator_test.go b/pkg/controller/nodeipam/ipam/range_allocator_test.go
index b13e66d7a1d..f697b581e8b 100644
--- a/pkg/controller/nodeipam/ipam/range_allocator_test.go
+++ b/pkg/controller/nodeipam/ipam/range_allocator_test.go
@@ -842,6 +842,15 @@ func TestReleaseCIDRSuccess(t *testing.T) {
        }
 }
 func TestNodeDeletionReleaseCIDR(t *testing.T) {
+       _, clusterCIDRv4, err := netutils.ParseCIDRSloppy("10.10.0.0/16")
+       if err != nil {
+               t.Fatalf("unexpected error: %v", err)
+       }
+
+       _, allocatedCIDR, err := netutils.ParseCIDRSloppy("10.10.0.0/24")
+       if err != nil {
+               t.Fatalf("unexpected error: %v", err)
+       }
        testCases := []struct {
                description       string
                nodeKey           string
@@ -857,7 +866,8 @@ func TestNodeDeletionReleaseCIDR(t *testing.T) {
                                                Name: "node0",
                                        },
                                        Spec: v1.NodeSpec{
-                                               PodCIDRs: []string{"10.10.0.1/24"},
+                                               PodCIDR:  allocatedCIDR.String(),
+                                               PodCIDRs: []string{allocatedCIDR.String()},
                                        },
                                },
                        },
@@ -873,7 +883,8 @@ func TestNodeDeletionReleaseCIDR(t *testing.T) {
                                                DeletionTimestamp: &metav1.Time{Time: time.Now()},
                                        },
                                        Spec: v1.NodeSpec{
-                                               PodCIDRs: []string{"10.10.0.1/24"},
+                                               PodCIDR:  allocatedCIDR.String(),
+                                               PodCIDRs: []string{allocatedCIDR.String()},
                                        },
                                },
                        },
@@ -887,15 +898,10 @@ func TestNodeDeletionReleaseCIDR(t *testing.T) {
                },
        }
 
-       clusterCIDR := "10.10.0.0/16"
-
        for _, tc := range testCases {
                t.Run(tc.description, func(t *testing.T) {
                        allocatorParams := CIDRAllocatorParams{
-                               ClusterCIDRs: func() []*net.IPNet {
-                                       _, clusterCIDRv4, _ := netutils.ParseCIDRSloppy(clusterCIDR)
-                                       return []*net.IPNet{clusterCIDRv4}
-                               }(),
+                               ClusterCIDRs:         []*net.IPNet{clusterCIDRv4},
                                ServiceCIDR:          nil,
                                SecondaryServiceCIDR: nil,
                                NodeCIDRMaskSizes:    []int{24},
@@ -908,7 +914,10 @@ func TestNodeDeletionReleaseCIDR(t *testing.T) {
                        _, tCtx := ktesting.NewTestContext(t)
 
                        fakeNodeInformer := test.FakeNodeInformer(fakeNodeHandler)
-                       nodeList, _ := fakeNodeHandler.List(tCtx, metav1.ListOptions{})
+                       nodeList, err := fakeNodeHandler.List(tCtx, metav1.ListOptions{})
+                       if err != nil {
+                               t.Fatalf("Failed to get list of nodes %v", err)
+                       }
                        allocator, err := NewCIDRRangeAllocator(tCtx, fakeNodeHandler, fakeNodeInformer, allocatorParams, nodeList)
                        if err != nil {
                                t.Fatalf("failed to create CIDRRangeAllocator: %v", err)
@@ -928,20 +937,19 @@ func TestNodeDeletionReleaseCIDR(t *testing.T) {
                                t.Fatalf("Expected a single cidrSet, but got %d", len(rangeAllocator.cidrSets))
                        }
 
-                       getNextCidr := rangeAllocator.cidrSets[0]
-                       nextCidr, _ := getNextCidr.AllocateNext()
-
-                       _, expectedNextCIDR, _ := netutils.ParseCIDRSloppy("10.10.0.0/24")
-
+                       // if the allocated CIDR was released we expect the nextAllocated CIDR to be the same
+                       nextCIDR, err := rangeAllocator.cidrSets[0].AllocateNext()
+                       if err != nil {
+                               t.Fatalf("unexpected error trying to allocate next CIDR: %v", err)
+                       }
+                       expectedCIDR := "10.10.1.0/24" // existing allocated is 10.0.0.0/24
                        if tc.shouldReleaseCIDR {
-                               if nextCidr.String() != expectedNextCIDR.String() {
-                                       t.Fatalf("Expected CIDR to be released, but it was not. Got: %v, Expected: %v", nextCidr, expectedNextCIDR)
-                               }
-                       } else {
-                               if nextCidr.String() == expectedNextCIDR.String() {
-                                       t.Fatalf("Expected CIDR not to be released, but it was. Got: %v", nextCidr)
-                               }
+                               expectedCIDR = allocatedCIDR.String() // if cidr was released we expect to reuse it
                        }
+                       if nextCIDR.String() != expectedCIDR {
+                               t.Fatalf("Expected CIDR %s to be allocated next, but got: %v", expectedCIDR, nextCIDR.String())
+                       }
+
                })
        }
 }
```

**Comment by adrianmoisey** [comment]
> @adrianmoisey I realized that we have to change something on the handlers too

> On the test I know that you took something from the existing ones, but it is better to always check all the errors, also I reorganized a bit the test to make it more clear how we are asserting

Thank you, that diff was super helpful, I've committed it and pushed.

**Comment by aojea** [comment]
/lgtm
/approve

Thanks

**Pull request merged**